### PR TITLE
Use proper validation context for branch parameters compilation

### PR DESCRIPTION
**Pull request by mproch**

### CODE REVIEW

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/PartSort.scala [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import pl.touk.nussknacker.engine.splittedgraph.SplittedNodesCollector
+import pl.touk.nussknacker.engine.splittedgraph.part.SourcePart
+import pl.touk.nussknacker.engine.splittedgraph.splittednode.EndingNode
+
+import scala.annotation.tailrec
+
+object PartSort {
+
+  @tailrec
+  //this is primitive topological sort, we want to sort parts,
```
Comment (arkadius) [inline engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/PartSort.scala]: Can we move it on the class level? And also write some details why this sorting is need? Or at least notice in the place of invocation...

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/ProcessCompiler.scala [hunk 2]
```diff
@@ -126,12 +129,23 @@ protected trait ProcessCompilerBase {
...
-      splittedProcess.sources.map(compile).sequence
+      compileSources(splittedProcess.sources)
     ) { (_, exceptionHandler, sources) =>
       CompiledProcessParts(splittedProcess.metaData, exceptionHandler, sources)
     }
   }
 
+  private def compileSources(sources: NonEmptyList[SourcePart])(implicit meta: MetaData): CompilationResult[NonEmptyList[PotentiallyStartPart]] = {
+    val zeroAcc = (CompilationResult(Valid(List[PotentiallyStartPart]())), BranchEndContexts(Map(), None))
+    val (result, _) = PartSort.sort(sources.toList).foldLeft(zeroAcc) { case ((resultSoFar, branchContexts), nextPart) =>
```
Comment (arkadius) [inline engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/ProcessCompiler.scala]: Can we put here some short notice why this sorting is need and reference to longer notice in PartSort? + maybe some TODO if it isn't final solution?

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/ProcessCompiler.scala [hunk 3]
```diff
@@ -126,12 +129,23 @@ protected trait ProcessCompilerBase {
...
+        val compiledPart = compile(nextPart, branchContexts)
+        //we don't use andThen on CompilationResult, since we don't want to stop if there are errors in part
+        val nextResult = CompilationResult.map2(resultSoFar, compiledPart)(_ :+ _)
+        (nextResult, branchContexts.addPart(nextPart.node, compiledPart))
```
Comment (arkadius) [inline engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/ProcessCompiler.scala]: I don't get it ... we add part also for non-join parts? Do we really need to do it in folding (next step depends on previous) or we can do it without such dependency: first some map and then sequence

File: engine/interpreter/src/test/scala/pl/touk/nussknacker/engine/compile/CustomNodeValidationSpec.scala [hunk 4]
```diff
@@ -257,6 +258,28 @@ class CustomNodeValidationSpec extends FunSuite with Matchers with OptionValues
...
+          .branch("join1", "unionTransformer", Some("outPutVar"),
+            List(
+              "branch1" -> List("key" -> "'key1'", "value" -> "#input")
+            )
+          )
+          .processorEnd("stringService", "stringService" , "stringParam" -> "''")
+      ))
+    val validationResult = validator.validate(process)
+
+    validationResult.variablesInNodes("stringService")("outPutVar") shouldBe TypedObjectTypingResult(Map("branch1" -> Typed[String]))
```
Comment (arkadius) [inline engine/interpreter/src/test/scala/pl/touk/nussknacker/engine/compile/CustomNodeValidationSpec.scala]: Can you add two branches with different value's types?

Comment (mproch) [reply engine/interpreter/src/test/scala/pl/touk/nussknacker/engine/compile/CustomNodeValidationSpec.scala]: added assertion in other test, this one is for different case

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/ProcessCompiler.scala (hunk 3 above)
Comment (mproch) [reply engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/ProcessCompiler.scala]: I added comment in code

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/ProcessCompiler.scala (hunk 2 above)
Comment (mproch) [reply engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/ProcessCompiler.scala]: Added comment

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/PartSort.scala (hunk 1 above)
Comment (mproch) [reply engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/PartSort.scala]: Added comment

**Pull request merged**